tools/worker-runner/run/state.go

```diff
@@ -98,3 +102,61 @@ func (state *State) CheckProviderResults() error {
 
 	return nil
 }
+
+// Write the state to the given cache file, checking permissions along the way
+func (state *State) WriteCacheFile(filename string) error {
+	log.Printf("Caching worker-runner state at %s", filename)
+	var encoded []byte
+	encoded, err := json.Marshal(&state)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(filename, encoded, 0700)
+	if err != nil {
+		return err
+	}
+
+	// This file contains secrets, so ensure that this is really only
+	// accessible to the file owner (and having just created the file, that
+	// should be the current user).
+	err = perms.MakePrivateToOwner(filename)
```

Comment:
Should we be doing this in a different order? Create empty file -> set permissions -> append content? This leaves it potentially readable for a bit and longer if we crash, right?

Reply:
Yeah, let's defer that to a followup though -- this is just moving existing code.  #3067.